

Exercise #1
--------------

Generic risk finding score calculus:

    1 low severity = 1
    
    1 medium severity = 10
    
    1 high severity = 40


Only not fixed values count towards this score.

API: https://api.probely.com/targets/RzXFSNHH3qUY/findings/ GET





Exercise #2
--------------

Calculate and compare the risk score of two scans:
    scan_id_1 = 3hbQvcGEmLbW
    scan_id_2 = 2RnxpEEm2qd5
    
    
API: https://api.probely.com/targets/RzXFSNHH3qUY/findings/?scan=3hbQvcGEmLbW GET


API: https://api.probely.com/targets/RzXFSNHH3qUY/findings/?scan=2RnxpEEm2qd5 GET
    
Compare two scans and print which findings were fixed (not present in the second scan),
which are still not fixed (appear in both scans) and which ones are new (only in the
second scan).
    

In [482]:
import requests
import pandas as pd
import numpy as np

In [483]:
id = []
url = []
severity = []
state = []
changed = []

In [484]:
def collect_api_data(results, scan_id):
    for result in results:
        
        id.append(result['id'])
        url.append(result['url'])
        if result['severity'] == 10:
            severity.append(1)
        elif result['severity'] == 20:
            severity.append(10)
        elif result['severity'] == 30:
            severity.append(40)
        if scan_id != '':
            state.append(result['state'])
            changed.append(result['changed'].split('T')[0])

            
        

In [485]:
def api_fetch_findings(target_id, scan_id):
    
    # Clear id, url, severity, state, changed lists before starting
    id.clear()
    url.clear()
    severity.clear()
    state.clear()
    changed.clear()
    
    # Check if has scan_id. If has scan_id perform capture on the findings of that particular scan
    if scan_id != '':
        api_endpoint = 'https://api.probely.com/targets/{}/findings/?scan={}'.format(target_id, scan_id)
    # If does not have a scan_id perform capture on the unfixed findings of that target
    else:
        api_endpoint = 'https://api.probely.com/targets/{}/findings/'.format(target_id)
    
    headers = (
            {
                'Content-Type': 'application/json',
                'Authorization': 'JWT eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0ZW5hbnQiOiJwcm9iZWx5IiwidXNlcm5hbWUiOiJZVWt3WjhHZFhpUmkiLCJqdGkiOiJRRDdoWUFvdjdTYnIifQ.O53R154sjyE0I5iv_ykFkboz7i5qeQwRRk-Kve9hjIs'
            })
    
    # GET all the data from the API
    page = requests.get(api_endpoint, headers=headers)
    # Capture the total ammount of pages, for any need of iteration between pages
    
    page_total = page.json()['page_total']
    
    # Capture results with json format
    results = page.json()['results']
    
    collect_api_data(results, scan_id)
    if page_total > 1:  
            for x in range(2, page_total+1):
                if scan_id != '':
                    api_endpoint = 'https://api.probely.com/targets/{}/findings/?scan={}&page={}'.format(target_id, scan_id, str(x))
                else:
                    api_endpoint = 'https://api.probely.com/targets/{}/findings/?page={}'.format(target_id, str(x))
                
                page = requests.get(api_endpoint, headers=headers)
                results = page.json()['results']
                collect_api_data(results, scan_id)
    
    # Create Dataframe to organize and visualize data on a better format
    if scan_id != '':
        dataframe = pd.DataFrame(data={'id': id, 'url': url, 'state': state, 'severity': severity, 'changed': changed})
    else:
        dataframe = pd.DataFrame(data={'id': id, 'url': url, 'severity': severity})
    print("The total ammount of cases is {}".format(dataframe.shape[0]))
    return dataframe


### Fetch API with empty 'scan_id' in 2nd argument in order to retrieve only the unfixed findings

In [486]:
target_id = "RzXFSNHH3qUY"
df = api_fetch_findings(target_id, '')

The total ammount of cases is 66


### Visualize the first 20 results (Number showing results, may be changed below)

In [487]:
df.head(20) # Show 20 results

id                                                url  severity
0    3                    https://test-0.ox.qa.prbly.win/        40
1   67          http://test-0.ox.qa.prbly.win/WackoPicko/        40
2   30  https://test-0.ox.qa.prbly.win/WackoPicko/gues...        40
3   22  https://test-0.ox.qa.prbly.win/WackoPicko/user...        40
4   18  https://test-0.ox.qa.prbly.win/WackoPicko/gues...        40
5   20  https://test-0.ox.qa.prbly.win/WackoPicko/user...        40
6   12         https://test-0.ox.qa.prbly.win/WackoPicko/        40
7   17  https://test-0.ox.qa.prbly.win/WackoPicko/pict...        40
8   13  https://test-0.ox.qa.prbly.win/dyn/goal/onClic...        40
9   23  https://test-0.ox.qa.prbly.win/WackoPicko/phpi...        40
10  25  https://test-0.ox.qa.prbly.win/WackoPicko/phpi...        40
11  27  https://test-0.ox.qa.prbly.win/WackoPicko/pass...        40
12  24  https://test-0.ox.qa.prbly.win/WackoPicko/phpi...        40
13  34  https://test-0.ox.qa.prbly.win/WackoPicko/cryp...        40
14  48         https://test-0.ox.qa.prbly.win/WackoPicko/        40
15  14  https://test-0.ox.qa.prbly.win/WackoPicko/picc...        40
16  50  https://test-0.ox.qa.prbly.win/WackoPicko/gues...        40
17  10                    https://test-0.ox.qa.prbly.win/        10
18   8                    https://test-0.ox.qa.prbly.win/        10
19  58  https://test-0.ox.qa.prbly.win/WackoPicko/gues...        10

In [488]:
severity_total_sum = df['severity'].sum() # The total sum of the severity for this target
severity_mean = int(df['severity'].mean()) # The mean||average of the severity for this target
low_severity_sum = (df['severity'] == 1).sum()
medium_severity_sum = (df['severity'] == 10).sum()
high_severity_sum = (df['severity'] == 40).sum()

In [489]:
print('The total score of unfixed cases is {}, with a mean of {} severity points.\nAmmount of low severity cases {}.  \
\nAmmount of medium severity cases {}. \
\nAmmount of high severity cases {}.'.format(severity_total_sum, severity_mean, low_severity_sum, medium_severity_sum, high_severity_sum))

The total score of unfixed cases is 927, with a mean of 14 severity points.
Ammount of low severity cases 27.  
Ammount of medium severity cases 22. 
Ammount of high severity cases 17.


## Conclusion from exercise 1:

    The total score of unfixed cases in the target_id: "RzXFSNHH3qUY" is 927, with a mean of 14 severity points.
    Ammount of low severity cases 24.  
    Ammount of medium severity cases 10. 
    Ammount of high severity cases 15.
    66 cases were captured in total.

In [490]:
scan_id_1 = '3hbQvcGEmLbW'
scan_id_2 = '2RnxpEEm2qd5'

df_1 = api_fetch_findings(target_id, scan_id_1)
df_2 = api_fetch_findings(target_id, scan_id_2)

The total ammount of cases is 49
The total ammount of cases is 50


### Change ['changed'] column into a datetime object

In [491]:
df_1['changed'] = pd.to_datetime(df_1['changed'])
df_2['changed'] = pd.to_datetime(df_2['changed'])

In [492]:
df_1_unfixed_cases = df_1.loc[df_1['state'] == 'notfixed']
total_ammount_unfixed_cases_1 = df_1_unfixed_cases.shape[0]

df_1_fixed_cases = df_1.loc[df_1['state'] == 'fixed']
total_ammount_fixed_cases_1 = df_1_fixed_cases.shape[0]

In [493]:
severity_total_sum = df_1['severity'].sum() # The total sum of the severity for this target
severity_mean = int(df_1['severity'].mean()) # The mean||average of the severity for this target
low_severity_sum = (df_1['severity'] == 1).sum()
medium_severity_sum = (df_1['severity'] == 10).sum()
high_severity_sum = (df_1['severity'] == 40).sum()

In [494]:
print('The total score of unfixed cases in the scan_id: "{}" is {}, with a mean of {} severity points.\nAmmount of low severity cases {}.  \
\nAmmount of medium severity cases {}. \
\nAmmount of high severity cases {}.'.format(scan_id_1, severity_total_sum, severity_mean, low_severity_sum, medium_severity_sum, high_severity_sum))

The total score of unfixed cases in the scan_id: "3hbQvcGEmLbW" is 724, with a mean of 14 severity points.
Ammount of low severity cases 24.  
Ammount of medium severity cases 10. 
Ammount of high severity cases 15.


### We can see below that on the second scan ('3hbQvcGEmLbW') we got 49 results in total, only 3 cases were completely fixed and 41 are still waiting to be fixed leaving 5 findings to review due to the possibility of false positives or false negatives.

In [495]:
df_1_unfixed_cases.head(10)

id                                                url     state  severity  \
0   3                    https://test-0.ox.qa.prbly.win/  notfixed        40   
1  30  https://test-0.ox.qa.prbly.win/WackoPicko/gues...  notfixed        40   
2  22  https://test-0.ox.qa.prbly.win/WackoPicko/user...  notfixed        40   
3  18  https://test-0.ox.qa.prbly.win/WackoPicko/gues...  notfixed        40   
4  20  https://test-0.ox.qa.prbly.win/WackoPicko/user...  notfixed        40   
5  12         https://test-0.ox.qa.prbly.win/WackoPicko/  notfixed        40   
6  17  https://test-0.ox.qa.prbly.win/WackoPicko/pict...  notfixed        40   
7  13  https://test-0.ox.qa.prbly.win/dyn/goal/onClic...  notfixed        40   
8  23  https://test-0.ox.qa.prbly.win/WackoPicko/phpi...  notfixed        40   
9  25  https://test-0.ox.qa.prbly.win/WackoPicko/phpi...  notfixed        40   

     changed  
0 2019-02-22  
1 2019-02-22  
2 2019-02-22  
3 2019-02-22  
4 2019-02-22  
5 2019-02-22  
6 2019-02-22  
7 2019-02-22  
8 2019-02-22  
9 2019-02-22

In [496]:
total_ammount_unfixed_cases_1

41

In [497]:
df_2_fixed_cases.head()

id                                                url     state  severity  \
14  14  https://test-0.ox.qa.prbly.win/WackoPicko/picc...   invalid        40   
23  40                https://test-0.ox.qa.prbly.win/foo/     fixed        10   
24  42                https://test-0.ox.qa.prbly.win/foo/     fixed        10   
47   7                    https://test-0.ox.qa.prbly.win/  accepted         1   

      changed  
14 2019-02-22  
23 2019-01-23  
24 2019-01-23  
47 2019-02-22

In [498]:
total_ammount_fixed_cases_1

3

In [499]:
df_1['changed'].value_counts()

2019-02-22    46
2019-01-23     2
2019-02-17     1
Name: changed, dtype: int64

In [500]:
df_2_unfixed_cases = df_2.loc[df_2['state'] == 'notfixed']
total_ammount_unfixed_cases_2 = df_2_unfixed_cases.shape[0]

df_2_fixed_cases = df_1.loc[df_2['state'] == 'fixed']
total_ammount_fixed_cases_2 = df_2_fixed_cases.shape[0]

In [501]:
severity_total_sum = df_2['severity'].sum() # The total sum of the severity for this target
severity_mean = int(df_2['severity'].mean()) # The mean||average of the severity for this target
low_severity_sum = (df_2['severity'] == 1).sum()
medium_severity_sum = (df_2['severity'] == 10).sum()
high_severity_sum = (df_2['severity'] == 40).sum()

In [502]:
print('The total score of unfixed cases in the scan_id: "{}" is {}, with a mean of {} severity points.\nAmmount of low severity cases {}.  \
\nAmmount of medium severity cases {}. \
\nAmmount of high severity cases {}.'.format(scan_id_2, severity_total_sum, severity_mean, low_severity_sum, medium_severity_sum, high_severity_sum))

The total score of unfixed cases in the scan_id: "2RnxpEEm2qd5" is 725, with a mean of 14 severity points.
Ammount of low severity cases 25.  
Ammount of medium severity cases 10. 
Ammount of high severity cases 15.


### We can see below that on the second scan ('2RnxpEEm2qd5') we got 50 results in total, only 4 cases were completely fixed and 41 are still waiting to be fixed leaving 6 findings to review due to the possibility of false positives or false negatives.

### The total risk score

In [503]:
df_2_unfixed_cases.head(10)

id                                                url     state  severity  \
0   3                    https://test-0.ox.qa.prbly.win/  notfixed        40   
1  22  https://test-0.ox.qa.prbly.win/WackoPicko/user...  notfixed        40   
2  18  https://test-0.ox.qa.prbly.win/WackoPicko/gues...  notfixed        40   
3  20  https://test-0.ox.qa.prbly.win/WackoPicko/user...  notfixed        40   
4  12         https://test-0.ox.qa.prbly.win/WackoPicko/  notfixed        40   
5  17  https://test-0.ox.qa.prbly.win/WackoPicko/pict...  notfixed        40   
6  13  https://test-0.ox.qa.prbly.win/dyn/goal/onClic...  notfixed        40   
7  23  https://test-0.ox.qa.prbly.win/WackoPicko/phpi...  notfixed        40   
8  25  https://test-0.ox.qa.prbly.win/WackoPicko/phpi...  notfixed        40   
9  27  https://test-0.ox.qa.prbly.win/WackoPicko/pass...  notfixed        40   

     changed  
0 2019-02-22  
1 2019-02-22  
2 2019-02-22  
3 2019-02-22  
4 2019-02-22  
5 2019-02-22  
6 2019-02-22  
7 2019-02-22  
8 2019-02-22  
9 2019-02-22

In [504]:
total_ammount_unfixed_cases_2

41

In [505]:
total_ammount_fixed_cases_2

4

### The findings from both scans are very similar in terms of dates:

    Both scans were changed 46 times in 2019-02-22
    
    Scan 3hbQvcGEmLbW was changed 2x in 2019-01-23 and 1x in 2019-02-17
    
    Scan 2RnxpEEm2qd5 was changed 2x in 2019-01-30 and 2x in 2019-02-17 
    

In [506]:
print('Findings from scan("3hbQvcGEmLbW")')
print(df_1['changed'].value_counts())
print('Findings from scan("2RnxpEEm2qd5")')
print(df_2['changed'].value_counts())

Findings from scan("3hbQvcGEmLbW")
2019-02-22    46
2019-01-23     2
2019-02-17     1
Name: changed, dtype: int64
Findings from scan("2RnxpEEm2qd5")
2019-02-22    46
2019-01-30     2
2019-02-17     2
Name: changed, dtype: int64


## Conclusion from exercise 2:

    The total score of unfixed cases in the scan_id: "3hbQvcGEmLbW" is 724, with a mean of 14 severity points.
    Ammount of low severity cases 24.  
    Ammount of medium severity cases 10. 
    Ammount of high severity cases 15.
    49 cases were captured in total, only 3 cases were completely fixed, 41 are still waiting to be fixed, leaving 5 findings to review due to the possibility of false positives or false negatives.
    Was changed 46x in 2019-02-22, 2x in 2019-01-23 and 1x in 2019-02-17
    
    
    
    The total score of unfixed cases in the scan_id: "2RnxpEEm2qd5" is 725, with a mean of 14 severity points.
    Ammount of low severity cases 25.  
    Ammount of medium severity cases 10. 
    Ammount of high severity cases 15.
    50  cases were captured in total, only 3 cases were completely fixed, 41 are still waiting to be fixed leaving 6 findings to review due to the possibility of false positives or false negatives.
    Was changed 46x in 2019-02-22, 2x in 2019-01-30 and 2x in 2019-02-17